src/include/processor/operator/generic_extend.h

```diff
@@ -35,44 +89,32 @@ class GenericExtend : public PhysicalOperator {
     bool getNextTuplesInternal() override;
 
     unique_ptr<PhysicalOperator> clone() override {
-        return make_unique<GenericExtend>(inVectorPos, outNodeVectorPos, adjColumnAndListCollection,
-            outPropertyVectorsPos, propertyColumnAndListCollections, children[0]->clone(), id,
-            paramsString);
+        unordered_map<table_id_t, unique_ptr<AdjAndPropertyCollection>> clonedCollections;
+        for (auto& [tableID, adjAndPropertyCollection] : adjAndPropertyCollectionPerNodeTable) {
+            clonedCollections.insert({tableID, adjAndPropertyCollection->clone()});
+        }
+        return make_unique<GenericExtend>(inVectorPos, outNodeVectorPos, outPropertyVectorsPos,
+            std::move(clonedCollections), children[0]->clone(), id, paramsString);
     }
 
 private:
-    bool scan();
-
-    inline bool hasColumnToScan() const {
-        return nextColumnIdx < adjColumnAndListCollection.columns.size();
-    }
-    inline bool hasListToScan() const {
-        return nextListIdx < adjColumnAndListCollection.lists.size();
-    }
-    bool scanColumns();
-    bool scanLists();
-    bool scanColumn(uint32_t idx);
-    bool scanList(uint32_t idx);
+    bool scanCurrentAdjAndPropertyCollection();
+    void initCurrentAdjAndPropertyCollection(const nodeID_t& nodeID);
 
 private:
+    // vector positions
     DataPos inVectorPos;
     DataPos outNodeVectorPos;
-    ColumnAndListCollection adjColumnAndListCollection;
     vector<DataPos> outPropertyVectorsPos;
-    vector<ColumnAndListCollection> propertyColumnAndListCollections;
-
+    // vectors
     shared_ptr<ValueVector> inVector;
     shared_ptr<ValueVector> outNodeVector;
     vector<shared_ptr<ValueVector>> outPropertyVectors;
-    unique_ptr<ListSyncState> listSyncState;
-    uint32_t nextColumnIdx;
-    uint32_t nextListIdx;
-    // A list may be scanned for multi getNext() call e.g. large list. So we track current list.
-    AdjLists* currentAdjList;
-    ListHandle* currentAdjListHandle;
-    vector<Lists*> currentPropertyLists;
-    vector<ListHandle*> currentPropertyListHandles;
-    node_offset_t currentNodeOffset;
+    // storage structures
+    unordered_map<table_id_t, unique_ptr<AdjAndPropertyCollection>>
+        adjAndPropertyCollectionPerNodeTable;
+    AdjAndPropertyCollection* currentAdjAndPropertyCollection = nullptr;
+    unique_ptr<ListSyncState> listSyncState = nullptr;
```

Comment:
You don't really need to keep a current list sync state here. I think it's more clear to move ListSyncState into AdjAndPropertyCollection, so each will have their own one.